### ISSUE TITLE
Question: selectively disable strict types for specific definition or file?

### ISSUE THREAD

**Issue by typhonrt** [ajv-validator/ajv#1401]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**
v7.0.3

**Ajv options object**
ajv-cli -> ajv compile --all-errors -c ajv-errors

I have the below JSON schema which I'd like to generate one error for the object type check for `authors+`. To accomplish this I omit `"type": "object"` in both the referenced definitions which just have a `properties` field. This is similar to the example at the bottom of the [strict-mode docs](https://github.com/ajv-validator/ajv/blob/master/docs/strict-mode.md#strict-types) except I'm use `$ref`.

The type in `authors+ items` applies and all works well except I do get logged warnings for `strictTypes` for the referenced definitions which is expected. Is there any way to selectively disable strictTypes just for these two definitions or perhaps even just a file? I've scoured the FAQ and searched and I gather the answer is no.  I couldn't get ajv-errors to create a single error either when `"type": "object"`is in the referenced definitions. So for the time being I have to disable strict types for the entire compilation step which I'd rather not do if possible. Conceivably this is an enhancement request for the ability to selectively disable strictTypes. Thanks for your work on `ajv`.

**JSON Schema**

```json
    "authors+": {
      "items": {
        "allOf": [
          { "$ref": "#/definitions/properties-author" },
          { "$ref": "#/definitions/properties-author+" }
        ],
        "type": "object",
        "title": "An object value"
      },
      "title": "An array of items",
      "type": "array"
    },
    "properties-author": {
      "properties" : {...}
    },
    "properties-author+": {
      "properties" : {...}
    }
```

**Comment by epoberezkin** [comment]
This is per-instance setting, so you can do it in different Ajv instance. Supporting options per schema would make caching compiled schemas more complex... I am not sure I understand why you can’t use “type” inside definitions? Do you want to pass validation in case it’s any other type?

**Comment by typhonrt** [comment]
@epoberezkin Thanks for your quick reply. Indeed I don't have full visibility on internal working of AJV and new to working w/ JSON schema and ajv though am an experienced developer. 

The reason for omitting the type in the properties-authors/+ definitions is that the authors+ version w/ two $refs will generate two duplicated type errors in the output vs the authors version with a single $ref produces a single type error. I have a full test suite to validate the generated validation functions and I'm trying to keep things as DRY as possible. Without the type: object in the properties the error output is the same for tests across the author / author+ variation and I don't have to duplicate tests. I'm generating schemas for 8 variations of similar manifest / JSON data along with associated validation functions. The two properties + a required warning turn into 20 warnings when executing via the ajv CLI / compile for omitting two "type" fields. The project I'm working on will eventually be moved to a community repo and while I'm certainly getting more acquainted with JSON schema that doesn't mean the next person will be as aware of potential errors that could be caught by strict mode of ajv. IE I can turn it off, but it might actually benefit the next developer potentially extending my initial work. Also turning on strict mode might confuse the next developer w/ all the warnings for an intentional / valid use of omitting the "type" attributes. The ideal solution from a usability side of things is being able to include a field in the schema that AJV understands to skip strict tests just for those sections, but understandably this just may not be feasible regarding the internal workings of AJV. No worries there. I was just curious about potential possibilities.

I do have a related question when it comes to "superfluous" keywords being listed in errors output of the validator functions. With some of the more comprehensive schema validation functions I'm creating I use if / else and other directives like oneOf, anyOf, allOf. At times these keywords will be in the error output along with the keywords catching the actual errors being validated. I gather some are pointing out potential problems with the schema itself or just adds data on what was processed.  An example (I consider the `if` keyword unnecessary):

```
[
   {
      "keyword": "required",
      "dataPath": "/media/0",
      "schemaPath": "#/definitions/media-item-strict/then/required",
      "params": {
         "missingProperty": "type"
      },
      "message": "should have required property 'type'"
   },
   {
      "keyword": "if",
      "dataPath": "/media/0",
      "schemaPath": "#/definitions/media-item-strict/if",
      "params": {
         "failingKeyword": "then"
      },
      "message": "should match \"then\" schema"
   }
]
```

My effort is to produce comprehensive validation tools for [Foundry VTT](https://foundryvtt.com/) manifests. Since many other 3rd party tooling related to this community may use the validation functions I'm producing I'd like to remove any unnecessary error output.  The only solution that I can think of is that I wrap the validation functions and filter out "superfluous" keywords as I couldn't find a way to have ajv suppress keywords that I'm not concerned about. I've scoured the docs, so maybe I've missed something as well. Is filtering the output the most appropriate route to take? IE I'm trying to produce validation functions that produce one error for one condition. 

This [wrapping / filtering](https://github.com/typhonjs-fvtt/manifest-plus/blob/main/dist/validators.js) solution does seen to do the trick. 

I plan to produce a separate module like [better-ajv-errors](https://www.npmjs.com/package/better-ajv-errors) to produce human readable error output w/ code frames. It looks like this particular project won't work w/ ajv v7+ due to usage of "jsonPointers" which was removed in v7.

As an aside, being open source work I'm actually streaming my coding and did a pretty good job of giving an overview of my effort at the beginning of this recent [stream](https://www.twitch.tv/videos/881667931). I struggled for ~8 hours at the end of the stream to wrap my head around the extra "superfluous" keywords. Again I'm new to AJV. Thanks for the input!
